### ISSUE TITLE
hevc_qsv encoder into ffmpeg, controlling options

### ISSUE THREAD

**Issue by eblanca** [Intel-Media-SDK/MediaSDK#2075]
Hi, I managed to build ffmpeg with MediaSDK support on my LMDE4 distro (it is Debian based) and when I ask for encoder options (in this case hevc_qsv) I get:
```
topolinik@debbieb1:~$ ~/ffmpeg -h encoder=hevc_qsv
ffmpeg version n4.2.2-6-gd1e8be3411 Copyright (c) 2000-2019 the FFmpeg developers
  built with gcc 8 (Debian 8.3.0-6)
  configuration: --prefix=/home/topolinik/svn/mpv-build/build_libs --enable-static --disable-shared --enable-gpl --disable-debug --disable-doc --enable-pic --enable-pic --disable-debug --enable-postproc --enable-gpl --enable-libdrm --enable-libtheora --enable-libxvid --enable-libx264 --enable-libvpx --enable-libmfx --extra-cflags=-I/opt/intel/mediasdk/include/ --extra-cxxflags=-I/opt/intel/mediasdk/include/ --extra-ldflags=-L/opt/intel/mediasdk/lib/ --enable-nonfree --enable-openssl
  libavutil      56. 31.100 / 56. 31.100
  libavcodec     58. 54.100 / 58. 54.100
  libavformat    58. 29.100 / 58. 29.100
  libavdevice    58.  8.100 / 58.  8.100
  libavfilter     7. 57.100 /  7. 57.100
  libswscale      5.  5.100 /  5.  5.100
  libswresample   3.  5.100 /  3.  5.100
  libpostproc    55.  5.100 / 55.  5.100
Encoder hevc_qsv [HEVC (Intel Quick Sync Video acceleration)]:
    General capabilities: delay hybrid 
    Threading capabilities: none
    Supported pixel formats: nv12 p010le qsv
hevc_qsv encoder AVOptions:
  -async_depth       <int>        E..V..... Maximum processing parallelism (from 1 to INT_MAX) (default 4)
  -avbr_accuracy     <int>        E..V..... Accuracy of the AVBR ratecontrol (from 0 to INT_MAX) (default 0)
  -avbr_convergence  <int>        E..V..... Convergence of the AVBR ratecontrol (from 0 to INT_MAX) (default 0)
  -preset            <int>        E..V..... (from 1 to 7) (default medium)
     veryfast                     E..V.....
     faster                       E..V.....
     fast                         E..V.....
     medium                       E..V.....
     slow                         E..V.....
     slower                       E..V.....
     veryslow                     E..V.....
  -rdo               <int>        E..V..... Enable rate distortion optimization (from -1 to 1) (default -1)
  -max_frame_size    <int>        E..V..... Maximum encoded frame size in bytes (from -1 to 65535) (default -1)
  -max_slice_size    <int>        E..V..... Maximum encoded slice size in bytes (from -1 to 65535) (default -1)
  -bitrate_limit     <int>        E..V..... Toggle bitrate limitations (from -1 to 1) (default -1)
  -mbbrc             <int>        E..V..... MB level bitrate control (from -1 to 1) (default -1)
  -extbrc            <int>        E..V..... Extended bitrate control (from -1 to 1) (default -1)
  -adaptive_i        <int>        E..V..... Adaptive I-frame placement (from -1 to 1) (default -1)
  -adaptive_b        <int>        E..V..... Adaptive B-frame placement (from -1 to 1) (default -1)
  -b_strategy        <int>        E..V..... Strategy to choose between I/P/B-frames (from -1 to 1) (default -1)
  -forced_idr        <boolean>    E..V..... Forcing I frames as IDR frames (default false)
  -low_power         <boolean>    E..V..... enable low power mode(experimental: many limitations by mfx version, BRC modes, etc.) (default false)
  -idr_interval      <int>        E..V..... Distance (in I-frames) between IDR frames (from -1 to INT_MAX) (default 0)
     begin_only                   E..V..... Output an IDR-frame only at the beginning of the stream
  -load_plugin       <int>        E..V..... A user plugin to load in an internal session (from 0 to 2) (default hevc_hw)
     none                         E..V.....
     hevc_sw                      E..V.....
     hevc_hw                      E..V.....
  -load_plugins      <string>     E..V..... A :-separate list of hexadecimal plugin UIDs to load in an internal session (default "")
  -profile           <int>        E..V..... (from 0 to INT_MAX) (default unknown)
     unknown                      E..V.....
     main                         E..V.....
     main10                       E..V.....
     mainsp                       E..V.....
  -gpb               <boolean>    E..V..... 1: GPB (generalized P/B frame); 0: regular P frame (default true)
```
Where can I get more details about them?
What I would like is more knowledge of what they do and how I can tune in order to get my personal balance for encoding time, visual quality and final file size but I need to understand them better.
Sorry if this is not the right place to ask for this.

**Comment by dmitryermilov** [comment]
Hi @eblanca ,

Detailed reply to your question requires a separate wiki page. Actually we've already considered such page but didn't start writing it. 
Right here, I can give you only some hints. I hope it will be enough. If not, please tell what exact option(s) you're interested in.
First of all, all parameters above are from MediaSDK API. You can find their description in the [manual](https://github.com/Intel-Media-SDK/MediaSDK/blob/master/doc/mediasdk-man.md). 
Let me describe just few (most important):
- -low_power - it enables a special type of HW encoders which use only fixed functions HW blocks (programmable hw execution units are not involved).  Can influence both performance and quality.
- -preset - it's for performance vs quality balancing. Pretty straightforward.
- -b_strategy - B frames gives higher quality but lower performance. 
- -async_depth - it influences performance, async_depth!=1 gives higher performance but means internal buffering (might be bad for low latency cases).

**Comment by eblanca** [comment]
Hello Dmitry, thanks for your hints.
While as a user I appreciate the idea of the wiki page, as a programmer I love that manual!
I am currently digging into ffmpeg source code and that manual has a lot of answers, great!
Gosh, I should have found it weeks ago!

I already had a brief view of the entire manual and some questions arose:

1. what's the difference in use case for the function classes `MFXVideoENCODE` and `MFXVideoENC`? (ffmpeg uses `MFXVideoENCODE` only)

2. in ffmpeg implementation, a call to `MFXVideoENCODE_Init` is issued passing the whole block of the user specified options, then before encoding starts, a call to `MFXVideoENCODE_GetVideoParam` passing an identical data block but empty. This is because this function returns the actual encoding parameter set due to the particular hardware/software (CPU/MSDK) configuration, is my understanding right?

3. the manual doesn't clarify this, but are any options usable only with particular codecs? For example, I see ffmpeg implementation provides lookahead options (lookahead, lookahead_depth and lookaheadDS) only for h264_qsv encoding engine, is this right (after reading the manual, I think it isn't) ?

Thank you.

**Comment by dmitryermilov** [comment]
Hello @eblanca ,

> what's the difference in use case for the function classes MFXVideoENCODE and MFXVideoENC? (ffmpeg uses MFXVideoENCODE only)

Most likely you don't need MFXVideoENC . MFXVideoENC is a special interface for encoder users who needs low level access to tune encoders quality. I'd advise to look at https://github.com/Intel-Media-SDK/MediaSDK/blob/master/doc/mediasdkfei-man.md. It's a FEI manual. FEI implements ENC/PAK interfaces.  This document explains idea of ENC and PAK (especially https://github.com/Intel-Media-SDK/MediaSDK/blob/master/doc/mediasdkfei-man.md#usage-models).

>This is because this function returns the actual encoding parameter set due to the particular hardware/software (CPU/MSDK) configuration, is my understanding right?

Exactly. Encoder implementation can internally initialize with slightly modified (e.g in case of some incompatibilities) version of mfxVideoParams passed at Init from application. So a user can check actual parameters by GetVideoParam.  

>the manual doesn't clarify this, but are any options usable only with particular codecs?

Yes. The manual is written in a lawyered language to some extend to be abstract from specific HW versions and a type of codec. 
Generally Query() API interface is provided to say if some API option is supported or not.  

>  For example, I see ffmpeg implementation provides lookahead options (lookahead, lookahead_depth and lookaheadDS) only for h264_qsv encoding engine, is this right (after reading the manual, I think it isn't)

In a nutshell, currently - yes, ffmpeg supports LookAhead only for AVC.
From MediaSDK perspective Lookahead can be also implemented for HEVC as well. We've recently proved it by https://github.com/Intel-Media-SDK/MediaSDK/pull/2034. Now the ball on ffmpeg side to enable such support for HEVC.

Regards,
Dmitry
